qemu/tests/pci_bridge.py

```diff
@@ -0,0 +1,286 @@
+    os_type = params["os_type"]
+    if os_type == "linux":
+        sub_test_type = params.get("sub_test_type", "dd_test")
+        for image in image_list:
+            params["dd_if"] = "ZERO"
+            params["dd_of"] = image
+            utils_test.run_virt_sub_test(test, params, env, sub_test_type)
+
+            params["dd_if"] = image
+            params["dd_of"] = "NULL"
+            utils_test.run_virt_sub_test(test, params, env, sub_test_type)
+    elif os_type == "windows":
+        cmd_timeout = int(params.get("cmd_timeout", 3600))
+        disk_letter = params.objects("disk_letter")
+        index_serialnumber = "wmic diskdrive get index,serialnumber"
+        disks = session.cmd_output(index_serialnumber, timeout=180)
+        disks = disks.splitlines()
+        if image_num > 3:
+            image_num = 3
+            fmt_list = ["virtio"] * image_num
+        else:
+            fmt_list = params.objects("disk_driver")
+        for num in range(image_num):
+            serial_num = "TARGET_DISK%d" % num
+            disk_index = None
+            for disk in disks:
+                if serial_num in disk:
+                    disk_index = disk.split()[0]
+            if disk_index is None:
+                test.error("Cannot find %s in guest!" % serial_num)
+            utils_misc.format_windows_disk(session, disk_index,
+                                           mountpoint=disk_letter[num])
+            iozone_cmd = params.get("iozone_cmd") % disk_letter[num]
+            iozone_cmd = utils_misc.set_winutils_letter(session, iozone_cmd)
+            status, output = session.cmd_status_output(iozone_cmd,
+                                                       timeout=cmd_timeout)
+            if status:
+                test.fail("Check '%s' block device '%s' failed! Output: %s"
+                          % (fmt_list[num], disk_letter[num], output))
+    else:
+        test.cancel("Unsupported OS type '%s'" % os_type)
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    [pci-bridge] Check devices which attached to 1 pci-bridge
+    [pci-bridge] Check devices which attached to different pci-bridges
+    [pci-bridge] Hotplug and unplug different devices to 1 pci-bridge
+    [pci-bridge] Hotplug and unplug different devices to different pci-bridges
+    This case will:
+    1) Attach one or multiple pci-bridge(s) to guest.
+    2) Prepare devices attached to pci-bridge(s).
+    3) Start the guest.
+    4) Hotplug devices (if needed).
+    5) Check devices.
+    6) Hotunplug devices (if needed).
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    error_context.context("Modify params!", logging.info)
+    pci_bridge_num = int(params.get("pci_bridge_num", 1))
+    prepare_pci_bridge(test, params, pci_bridge_num)
+    pci_bridges = params.objects("pci_controllers")
+
+    device_num = 0
```

Comment:
Seems unnecessary.  You have default value in prepare_images(device_num=0). And I don't see anywhere else pass this parameter, guess prepare_images() does not really need this parameter, only need it be returned.